File: src/Vpiska.Api/Vpiska.Api.Rest/Filters/ExceptionFilter.cs
using System.Threading.Tasks;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.Mvc.Filters;
using Serilog;
using Vpiska.Application;
using Vpiska.Domain.User;

namespace Vpiska.Api.Rest.Filters
{
    public sealed class ExceptionFilter : IAsyncExceptionFilter
    {
        private readonly ILogger _logger;
        
        public ExceptionFilter(ILogger logger)
        {
            _logger = logger;
        }

        public Task OnExceptionAsync(ExceptionContext context)
        {
            _logger.Error(context.Exception, "Unknown error");

            context.Result =
                new ObjectResult(new ApiResponse(false, null,
                    new[] { new ErrorResponse(Errors.mapAppError(AppError.InternalError)) })) { StatusCode = 200 };
            
            context.ExceptionHandled = true;
            return Task.CompletedTask;
        }
    }
}

File: src/Vpiska.Api/Vpiska.Api.Control/Filters/ExceptionFilter.fs
namespace Vpiska.Api.Control.Filters

open System.Threading.Tasks
open Microsoft.AspNetCore.Mvc
open Microsoft.AspNetCore.Mvc.Filters
open Serilog
open Vpiska.Api.Control.Http
open Vpiska.Domain.User

type ExceptionFilter(logger: ILogger) =
    
    interface IAsyncExceptionFilter with
        member this.OnExceptionAsync(context) =
            logger.Error(context.Exception, "Unknown error")
            context.Result <- ObjectResult({ IsSuccess = false
                                             Result = null
                                             Errors = [| { ErrorCode = AppError.InternalError |> Errors.mapAppError } |] }, StatusCode = 200)
            context.ExceptionHandled <- true
            Task.CompletedTask
